### ISSUE TITLE
editor.action.sortLinesAscending has confusing sort order for symbols

### ISSUE THREAD

**Issue by ayrtonmassey** [Microsoft/vscode#48123]
When sorting lines using `editor.action.sortLinesAscending` ("Sort Lines Ascending" via the Command Palette) the lines are sorted in a very strange order when the comparison involves a symbol character.

I would expect that `.` (period) would be sorted before `_` (underscore) because in ASCII, period is 46 and underscore is 95.

However, vscode uses `localeCompare` as its sorting method (see [sortLinesCommand.ts:76]( https://github.com/Microsoft/vscode/blob/54e7055c12f4e9a80f44c67758a12cf248d5f374/src/vs/editor/contrib/linesOperations/sortLinesCommand.ts#L76)), which results in underscore being sorted before period. See the repro steps below for an example.

I saw a previous issue regarding this (#15516) but it was closed because the ASCII ordering was in fact correct. In the case I've outlined below, the ordering is not correct.

I'm not sure what the correct solution would be but I think that for ASCII symbols, ASCII ordering should be obeyed. I do realise that "Sort Lines Ascending" is an ambiguous term - ascending according to what criteria? - so perhaps the command could be renamed to something more specific, or you could provide different default sorting options.

- VSCode Version: 1.22.1 Commit 950b8b0d37a
- OS Version: Windows 10 Enterprise Version 1709 OS Build 16299.371

Steps to Reproduce:

1. Copy the following into a new file:

```
a_b.txt
a_b_c.txt
````

2. Highlight the file contents.

3. Open the Command Palette and select "Sort Lines Ascending".

**Expected:**

The file is sorted in the following order:

```
a_b.txt
a_b_c.txt
```

**Actual:**

The file is sorted in the following order:

```
a_b_c.txt
a_b.txt
```

Does this issue occur when all extensions are disabled?: Yes

**Comment by alexandrudima** [comment]
I don't have a strong opinion either way. However, the current usage of `localeCompare` is consistent to our using `localeCompare` in other places, like e.g. in the explorer:

![image](https://user-images.githubusercontent.com/5047891/38945820-db766ef8-4337-11e8-9ee2-8349117b4c1e.png)

**Comment by ayrtonmassey** [comment]
Thanks for your response - interesting that the editor uses the same sort as the UI!

I use the sort function most often to sort lines in code, which usually contain only ASCII characters. The issue I (personally) have with the current vscode sort function is that when I pipe the file content into other utilities, those utilities expect lines to be in ASCII order.

Regarding the sorting of files in vscode's explorer, Windows explorer sorts the files as I would expect:

![image](https://user-images.githubusercontent.com/700278/38946992-6bca9490-433b-11e8-90e8-8b4b7accc8d1.png)

Sorting in ASCII order is beneficial as shorter filenames are placed before longer ones.

Since this is supposed to be a general sort & is also applied to UI elements, it's difficult to come up with a one-size-fits-all solution. Personally, I would introduce two settings:

-  `sorting.method` (or similar) which allows the options:
    - `ascii`, for sorting code / lists of files.
    - `localeCompare`, for sorting general text which may contain unicode characters, etc.
- `sorting.caseSensitivity` (or similar) which allows the options:
    - `caseInsensitive`, values converted to lowercase before comparison
    - `caseSensitive`, values compared without modification

Perhaps these settings could be separate from the sorting in the explorer, e.g. `editor.sorting.method` vs. `explorer.sorting.method`.

**Comment by benkimball** [comment]
I'm currently sorting a very long YAML file in order for it to pass [yamllint](https://github.com/adrienverge/yamllint) configured to enforce lexical sorting of keys. This issue is of interest to me because currently Code's idea of a lexical sort and `yamllint`'s idea differ in the area of underscores:

After Code `editor.action.sortLinesAscending`:

```yaml
---
underscore_underscore: second
underscore: first
```

yamllint output:

    3:1       error    wrong ordering of key "underscore" in mapping  (key-ordering)

**Comment by sathya-beeline** [comment]
Yeah. I find that the `sortLinesAscending` function is un-usable for me since it **does not** follow lexicographic ordering based on string length.

For example:

```ruby
gem 'graphql-batch', '~> 0.3.9'
gem 'graphql-errors', '~> 0.1.0'
gem 'graphql-preload', '~> 2.0.1'
gem 'graphql', '~> 1.8'
```

with lexicographic ordering should instead be

```ruby
gem 'graphql', '~> 1.8'
gem 'graphql-batch', '~> 0.3.9'
gem 'graphql-errors', '~> 0.1.0'
gem 'graphql-preload', '~> 2.0.1'
```

**Comment by sathya-beeline** [comment]
Yeah, I've run across this issue with the sort method as well.

One problem is that if you use unicode integer representations to sort, it will always be confusing. And you sort of have to use unicode to sort, since 

For example: `'` is greater than `-` according to VSCode.

So, the following would be in VSCode sort order:

```
gem 'graphql-example'
gem 'graphql'
```

And so would the following:

```
graphql-example
graphql
```

I believe this is because it uses the `SMALL HYPHEN-MINUS` character, and compares that to one of the single apostrophes, which have a higher integer representation.

Citation: https://www.ssec.wisc.edu/~tomw/java/unicode.html

**Comment by hamstergene** [comment]
I find it wrong that despite 99.999% of files in the world of coding use only ASCII in names and content, VSCode defaults to sorting in an unexpected way just so 0.001% of unicode cases could be satisfied. Unicode collation is for natural languages but programmer's editor/IDE more often than not deals with computer languages so is defaulting to `localeCompare` really justified?

I wonder if there can be a third option that does not involve switching editor setting back and forth? For example, do ascii sort first then do stable `localeCompare` sort with `ignorePunctuation: true`.

**Comment by CristianOspina** [comment]
For someone looking to fix this in the Explorer, this config can help:
"explorer.sortOrder": "type"

**Comment by rcdailey** [comment]
Three years and 17 upvotes later and not a peep from the developers. I get that there's a lot going on, but this can't really be that difficult to address, right?

**Comment by romain20100** [comment]
@rcdailey There are 5K issues at the moment, and this one is far from being critical.
Plus 17 upvotes doesn't quite compare with the millions of people using vscode.
Either way this comment won't help anyway. 🤷 
---

Just wanted to mention that I would like to be able to sort lines in "unicode" mode (similar to how explorer files are sorted, with the relevant option).
I didn't find any option to do that for the sort-lines command.

Could this setting be added? Has it been discussed already?
I did not find an issue or pr addressing this.

**Comment by jensbodal** [comment]
Interesting that a similar issue was closed 6 years ago since this is "as designed": https://github.com/microsoft/vscode/issues/15516

They used this example

```
['margin: 0;', 'margin-bottom: auto;'].sort();
// results in
["margin-bottom: auto;", "margin: 0;"]
```

However 

```
['react-foo', 'react', 'apple'].sort()
// results in
['apple', 'react', 'react-foo']

// vscode sort lines ascending results in
[
  'apple', 'react-foo', 'react', 
]
```

**Comment by brewster1134** [comment]
similar issues... with what i write as correctly sorted...

```
A: 'a',
A_B: 'ab'
```

when it is auto sorted, it switches the order

```
A_B: 'ab'
A: 'a',
```

i want to keep the shorter version first, with the following with additional granularity after, im sure this has been discussed before, but was unable to find anything. i have played around with various `eslintrc` settings, but no luck

i get `Expected object keys to be in natural insensitive ascending order` error

**Comment by sshishov** [comment]
Can we have at least the ability to override the sorting in the **settings**? It would be solve the issue almost for 99% of people.

I am pretty sure a lot of people using built in sorting abilities of IDE and this issue is frustrating. We also have this issue regarding `dot (.)` and `underscore (_)`

**Comment by roshal** [comment]
frustrating behavior for `editor.action.sortLines`

```
Z
z
A
a
```

```
a
A
z
Z
```

encouraging behavior for gnu sort

```sh
sort <<< "
Z
z
A
a
"
```

```
A
Z
a
z
```

**Comment by roshal** [comment]
encouraging behavior for gnu sort

```sh
sort <<< "
Z
z
A
a
"
```

```
A
Z
a
z
```

**Comment by roshal** [comment]
https://github.com/microsoft/vscode/pull/97272

`"explorer.sortOrderLexicographicOptions": "unicode",`

it would be nice to have the same option for the editor

`"editor.sortOrderLexicographicOptions": "unicode",`
